### DISCUSSION TITLE
Data generation for detectors

### DISCUSSION THREAD

**Discussion by PhilClemson** [dstl/Stone-Soup#481] [Q&A] [answered]
Hi All,

I'm currently preparing a pull request which will add some basic detectors (beamformers) for sonar data. This work is an output of Project BLUE, for which there was a "Show and Tell" event earlier this week that some may have attended.

I've been following the contributing guide and have a question regarding the tests.

The code I currently have reads directly from a pre-generated csv file (containing either externally measured or simulated data) and transforms this to a set of detections. However, it seems that the standard within Stone-Soup is to generate test data internally. So far I have only seen this done for detections / target tracks - I'm not sure what the equivalent should be when testing a detector method with the function of generating these detections.

My guess is that I need to include some simulator method to generate the raw sonar data so that it is possible to run tests and tutorials / demonstrations within Stone-Soup with out dependance on externally-generated data.

Is this right or would anyone advise a different approach?

Thanks!

**Comment by apiszcz** [comment]
I received guidance from the developers here 
1. Create a class 
 class NRTDetectionReader(DetectionReader):
2. Create a detections_gen method that reads your CSV data and returns it with a yield. 
     @BufferedGenerator.generator_method
    def detections_gen(self):
      read your data and
      yield timestamp, detectiondataset

I use pygtaild to monitor the end of the CSV file, it is pretty flexible and allows checkpointing and read from start, etc.

There may be more details on the Q&A in the gitter channel https://gitter.im/dstl/Stone-Soup

@sdhiscocks will probably reply with more insight.

**Reply by PhilClemson** [reply]
Thats good to know thanks. Fortunately that's pretty much exactly what I've done. I am using loadtxt() from numpy rather than pygtaild to import the data but hopefully that's okay?

**Comment by sdhiscocks** [comment, marked as answer]
I think for tests it may be best to take a small subsample of your sensor data (simulated or measured; enough to show detector working as expected) and put it in alongside the test script in the test folder.

You should be able to get the path to the file with something like: `os.path.join(os.path.dirname(__file__), "my_test_file.txt")`

**Reply by PhilClemson** [reply]
Thanks, that sounds easier than implementing a simulator in Stone-Soup (although that's something which could definitely be added at a later stage).

Looking forward to getting feedback once the pull request is submitted!
